src/Model/Table/SentencesTable.php

```diff
@@ -944,7 +944,28 @@ public function saveNewSentence($text, $lang, $userId, $correctness = 0, $basedO
             return false;
         }
 
-        return $this->saveWithDuplicateCheck($newSentence);
+        $sentence = $this->find('all')
+        ->where(['text' => $newSentence->text, 'lang' => $lang])
+        ->first();
+
+        // Duplicate sentence found
+        if ($sentence != null) {
```

Comment:
Using a hash was a suggestion from sysko which I simply followed when implementing the duplicate detection. I wouldn't be able to elaborate on why exactly, in all honesty, I simply trusted him that it would most likely be one of the more optimal solutions to detect duplicates.

Just for the record, the hash mechanism was first implemented to avoid duplicates for the vocabulary feature (#1218). It was later on replicated into the sentences by another developer (#577).

> So maybe we should stop using a hash and rely on the database for uniqueness of a sentence? It should be at least as efficient (or even more efficient)

If it simplifies the code and is just as efficient, then yes, we could do that.

Reply:
Sorry for the late answer but I wanted to get some more data.

> Using a hash was a suggestion from sysko which I simply followed when implementing the duplicate detection. I wouldn't be able to elaborate on why exactly, in all honesty, I simply trusted him that it would most likely be one of the more optimal solutions to detect duplicates.

I guess that at that time you couldn't use the full `text` field as an index and thus he thought that using a hash would solve the problem. The name of the index in the database (`dedup_idx`) suggests that this index was added for Horus which uses its own API for accessing the database.

But after reading and thinking more about this problem, I think even using a shorter prefix index (e.g. `text(100)`) would have helped at that time because the main reason for using an index is to reduce the number of rows the database needs to search for the wanted row. And just using the first e.g. 100 bytes reduces the number of candidates to 1 row in almost all cases[*]. That's the same situation we have with the hash collisions because the length of the hash is also limited.

[*] I've searched the longest common prefixes in the whole corpus (including the unreliable sentences) and there are 1815 pairs with a common prefix of more than 100 bytes, 360 of them have more than 150 bytes in common and 124 have more than 200 bytes in common. (You can download all common prefixes greater than 50 bytes [here](https://downloads.tatoeba.org/index_test/common-prefix-50.bz2)).

I've also tried to determine how much space we could save with a shorter prefix. On the dev server I've changed the prefix length, run `optimize table sentences` afterwards and checked the index length with `select index_length from information_schema.tables where table_name = 'sentences'`. I got the following values:

prefix length|index length (in bytes)
-|-
10|1,276,887,040
20|1,360,904,192
100|1,547,550,720
200|1,556,987,904
767 (current setting)|1,559,085,056
no prefix (full column)|1,559,085,056

As you can see there is no difference between the current setting and the full column (I guess internally MariaDB also limits the length of the index hash) and the difference to a prefix 100 is relatively small (10MB). Only going down to a very small prefix would get us some space reduction. But all in all I think a index for the full column is more useful than one with a very short prefix and so I would opt for the full one.

Finally I've tried to measure the speed difference between the old version using `DuplicateBehavior` and the version suggested by HelioStrike using the database itself. Therefore I added some timing calls to the code and modified the method to randomly use one of the two versions (see https://github.com/Tatoeba/tatoeba2/compare/dev...AndiPersti:profiling). On my local machine I run the testsuite for the `SentencesTable` three times and on the dev server I've used three different index settings (full column, prefix 767, prefix 20) and added manually random sentences (duplicates and new ones; we don't have `phpunit` installed on the dev server). The average times (in ns) I got were (raw data [here](https://downloads.tatoeba.org/index_test/)):

Setting | `DuplicateBehavior` | Database
-|-|-
local | 2,337,677 | 2,354,544
dev, prefix 20 | 1,699,825 | 1,289,747
dev, prefix 767|1,472,743 |1,345,367
dev, full column|1,563,595|1,527,053

It looks like the current version is slightly slower.

> If it simplifies the code and is just as efficient, then yes, we could do that.

It will definitely simplify the code and taking all the above into account I'd say we should change the way we look for duplicates.